### DISCUSSION TITLE
Renovate failing to look up private go module

### DISCUSSION THREAD

**Discussion by kchan623** [renovatebot/renovate#21386] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Renovate can look up private dependencies in go.mod

### What you tried so far.

Hello, I'm testing out Renovate for the first time and have tried configuring different `hostRules` with different `matchHost` values. I keep getting for one particular dependency from a private Go module: `Failed to look up go package github.com/<org>/<repo> (github.com/<org>/<repo>)(packageFile="go.mod", dependency="github.com/<org>/<repo>")`. I've verified that my GH PAT works since Renovate was able to fetch a dependency update for Terraform from another repository in the same GH organization. These two repositories are all under the same private organization as the repository I'm running Renovate in. I've tried looking up similar discussions/issues, and tried suggestions from there by changing `hostRules` config to no avail. Can someone help point out what I'm missing. Thank you!

At one point I was getting an additional error:
```
DEBUG: GET https://proxy.golang.org/github.com/<org>/<repo>/@v/list = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=1464)
DEBUG: Goproxy error: trying next URL provided with GOPROXY
{
  "err": {
    "message": "Could not resolve to a Repository with the name '<org>/<repo>'.",
    "stack": "Error: Could not resolve to a Repository with the name '<org>/<repo>'.\n    at GithubGraphqlDatasourceFetcher.doRawQuery (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/github/graphql/datasource-fetcher.js:84:29)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async GithubGraphqlDatasourceFetcher.doShrinkableQuery (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/github/graphql/datasource-fetcher.js:135:26)\n    at async GithubGraphqlDatasourceFetcher.doPaginatedQuery (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/github/graphql/datasource-fetcher.js:167:33)\n    at async GithubGraphqlDatasourceFetcher.getItems (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/github/graphql/datasource-fetcher.js:206:21)\n    at async Function.query (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/github/graphql/datasource-fetcher.js:31:23)\n    at async queryTags (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/github/graphql/index.js:8:17)\n    at async GithubTagsDatasource.getReleases (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/datasource/github-tags/index.js:48:28)\n    at async GoDirectDatasource.getReleases (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/datasource/go/releases-direct.js:54:23)\n    at async /home/ubuntu/renovateapp/node_modules/renovate/dist/util/cache/package/decorator.js:69:24\n    at async GoProxyDatasource.getReleases (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/datasource/go/releases-goproxy.js:50:30)\n    at async /home/ubuntu/renovateapp/node_modules/renovate/dist/util/cache/package/decorator.js:69:24\n    at async /home/ubuntu/renovateapp/node_modules/renovate/dist/util/cache/package/decorator.js:69:24\n    at async fetchReleases (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/datasource/index.js:256:19)\n    at async getPkgReleases (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/datasource/index.js:300:34)\n    at async lookupUpdates (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/lookup/index.js:52:45)\n    at async fetchDepUpdates (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/fetch.js:56:25)\n    at async /home/ubuntu/renovateapp/node_modules/p-map/index.js:57:22"
  }
}
DEBUG: Failed to look up go package github.com/<org>/<repo> (github.com/<org>/<repo>)(packageFile="go.mod", dependency="github.com/<org>/<repo>")
```

That error has gone away with my latest renovate config here:
```
"hostRules": [
    {
      "matchHost": "api.github.com",
      "hostType": "github",
      "encrypted": {
        "token": "<encrypted_token>"
      }
    },
    {
      "matchHost": "https://api.github.com/repos/<org>/",
      "encrypted": {
        "token": "<encrypted_token>"
      }
    },
```

Now all I'm getting is:
```
Failed to look up go package github.com/<org>/<repo> (github.com/<org>/<repo>)(packageFile="go.mod", dependency="github.com/<org>/<repo>")
```

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by rarkins** [comment]
If they're on the same org and both private then you shouldn't need any additional hostRules for it to work - the default GitHub token for the job should have access

**Comment by kchan623** [comment]
Hello! With the additional `hostRules` removed, I still get `DEBUG: Failed to look up go package github.com/<org>/<repo> (github.com/<org>/<repo>)(packageFile="go.mod", dependency="github.com/<org>/<repo>")`. To get the terraform dependency working on the other repository, I did have to specifically add this hostRule:
```
{
            "matchHost": "https://api.github.com/repos/<org>/<other-private-tf-repo>",
            "hostType": "github-tags",
            "encrypted": {
                "token": "encrypted_token"
            }
        },
```

When I added the Renovate GH App to the GH organization, I only enabled it to select repositories, one of them being the repository I'm running Renovate on. The Renovate App does not have access to the other repositories where the Go and Terraform dependencies are coming from. So that's why I thought to add the additional `hostRules`. But I'm not sure if this same reason is also why it's still having access issues? And I'm also not sure why adding the `hostRules` works for getting the Terraform updates from the other repository but not the repository containing the Go module.

**Reply by rarkins** [reply]
You need to add renovate to all of the repositories if you want the same token to work.

**Reply by kchan623** [reply]
Thank you for following up! So is it not possible at all to try to do the scenario described above? E.g. Not give Renovate App access to the dependency repository itself but pass a GH PAT that has all repo access within the organization in hostRule definitions? 
And if it's not possible, I'm still a bit confused why Renovate was able to access the other repository containing the TF dependency using the hostRule even though I also did not explicitly grant it access through the GH App config?

**Reply by rarkins** [reply]
It's possible, but I don't recommend it due to the challenges of getting it working

**Reply by kchan623** [reply]
I see, that makes sense. Thank you for following up!

**Reply by rarkins** [reply]
If you really can't install Renovate in the upstream repos, be sure to check out the examples in https://docs.renovatebot.com/getting-started/private-packages/#example-for-gomod

**Reply by rarkins** [reply]
You probably would need to match the PAT against the _specific repo_ though, not org. If you match it against the org, it can conflict with the built-in token Renovate uses automatically
